Add support for negative hexadecimal numbers

wismill opened the pull request:
Current implementation segfaults on negative numbers. Fix this issue by supporting them.

Note: While I think the generic implementation works, I am not so confident about the rewrite rules & inlinable annotations.

May require some update to the doc in #10.

Bodigrim reviewed (commented):
Usually when you wish to express a signed value via hexadecimal representation, the latter is written unsigned even if the value is negative. E. g, `0xff :: Int8`, never `-0x01 :: Int8` or `0x-01 :: Int8`.

wismill commented:
You mean to rely on the overflow? I am not sure this is a good idea. E.g. GHC will gladly accept `-0x01 :: Int8` but will warn on `0xff :: Int8`, and rightly so.

But I see this is the behaviour of `printf`. Also, it does not allow to format as `-0x01`.

I will give it a try.

wismill commented:
Done: 
```haskell
runBuffer (\b -> b |>& (-1 :: Int8)) == "ff"
```

Bodigrim reviewed (commented):
I'm AFK and cannot review properly, back on Monday. That said, I'm somewhat surprised at the scale of the changes. I'd imagine the existing code mostly work for negatives as well, except `shiftR` to be replaced by non-sign-extending version if `Data.Bits. isSigned`.

wismill commented:
The idea is to create RULES to use the same implementation for `Word`s and `Int`s; `Int`s are casted to their corresponding `Word`s.

But for other types, it is not clear how to optimize this. So I adapted the code from the `printf` function in `base` to handle those types using `Integer`.

wismill commented:
Rebased on #15. I added the tests and benchmarks of the previous implementation of this PR.

I also optimised `dropNibble` for common integer sizes. I left a commit with an alternative optimization using rules for discussion only.

When this PR is ready, all commits should be squashed.

Bodigrim commented:
LGTM, waiting for you to mark as ready.

If you do not mind, I'd prefer to merge unsquashed history here.

wismill commented:
> If you do not mind, I'd prefer to merge unsquashed history here.

I do not mind, if you think this is useful. I wonder if we should add tests using non-`base` integers such as in  [`wide-word`](https://hackage.haskell.org/package/wide-word), for completeness.

Bodigrim commented:
You can test `Int30`, but let's avoid additional dependencies.

wismill commented:
I added a test for `Int30`. It required fixing `Int30`, to handle negative values. This the same fix as in #17.

wismill commented:
@Bodigrim I made a mistake with `Int30` and it looks the CI is looping, but I have no way to cancel it 😅

wismill commented:
Fixed. Implementation seems correct, but there may be optimization opportunities.

I deactivated test for Decimal with `Int30`, since they now need the fix in #17.

wismill commented:
Note to self: ensure coverage using general `IntN n` from #17.

wismill commented:
Rebased. The test now uses random _signed_ integer sizes in range (8, 64). We could use the same approach with _unsinged_ integers. My experiments with tests should be squashed; I left them for now in case we need them for discussion.

Bodigrim commented on src/Data/Text/Builder/Linear/Hex.hs:
```diff
@@ -9,33 +9,48 @@ module Data.Text.Builder.Linear.Hex (
 buffer |>& n =
   appendBounded
-    (finiteBitSize n `shiftR` 2)
+    (1 + ((finiteBitSize n - 1) `shiftR` 2))
```
Let's move this line into a helper function, with `{-# INLINEABLE #-}` pragma.

wismill replied:
Done

Bodigrim commented:
Looks good, please rebase and rearrange commits as you wish them merged.

wismill commented:
@Bodigrim rebased. I reworked the commit history, but I kept the attempt with rewrite rules as your required.

I also added tests for <em>un</em>signed  numbers with random sizes. They should probably added to decimal builder as well, but let’s keep focus.

Bodigrim commented:
Great stuff!
